src/network/local_socket_reply.cpp

```diff
@@ -0,0 +1,232 @@
+    default:
+        if (httpStatusCode > 500)
+        {
+            // some kind of server error
+            code = QNetworkReply::UnknownServerError;
+        }
+        else
+        {
+            // content error we did not handle above
+            code = QNetworkReply::UnknownContentError;
+        }
+    }
+
+    return code;
+}
+} // namespace
+
+mp::LocalSocketReply::LocalSocketReply(const QString& socket_path, const QNetworkRequest& request,
+                                       QIODevice* outgoingData)
+    : QNetworkReply(), local_socket{std::make_unique<QLocalSocket>(this)}, reply_data{QByteArray(len, '\0')}
+{
+    QIODevice::open(QIODevice::ReadOnly);
+
+    local_socket->connectToServer(socket_path);
+
+    if (!local_socket->waitForConnected(5000))
+    {
+        throw std::runtime_error(fmt::format("Cannot connect to {}: {}", socket_path, local_socket->error()));
+    }
+
+    QObject::connect(local_socket.get(), &QLocalSocket::readyRead, this, &LocalSocketReply::read_reply);
+
+    send_request(request, outgoingData);
+}
+
+mp::LocalSocketReply::~LocalSocketReply()
+{
+    local_socket->disconnectFromServer();
+}
+
+void mp::LocalSocketReply::abort()
+{
+}
+
+qint64 mp::LocalSocketReply::readData(char* data, qint64 maxSize)
+{
+    if (offset < content_data.size())
+    {
+        qint64 number = qMin(maxSize, content_data.size() - offset);
+        memcpy(data, content_data.constData() + offset, number);
+        offset += number;
+
+        return number;
+    }
+
+    return -1;
+}
+
+void mp::LocalSocketReply::send_request(const QNetworkRequest& request, QIODevice* outgoingData)
+{
+    QByteArray http_data;
+    http_data.reserve(1024);
+
+    auto op = request.attribute(QNetworkRequest::CustomVerbAttribute).toByteArray();
+
+    // Build the HTTP method part
+    http_data += op;
+    http_data += ' ';
+    http_data += request.url().path();
+    http_data += " HTTP/1.1\r\n";
+
+    // Build the HTTP Host header
+    // Host can be anything, so we'll use 'multipass'
+    http_data += "Host: multipass\r\n";
+
+    // Build the HTTP User-Agent header
+    // We'll just use what Qt uses
+    http_data += "User-Agent: Mozilla/5.0\r\n";
+
+    // Build the HTTP Accept header
+    // Default to accept everything
+    http_data += "Accept: */*\r\n";
+
+    if (op == "POST" || op == "PUT")
+    {
+        http_data += "Content-Type: application/x-www-form-urlencoded\r\n";
+
+        if (outgoingData)
+        {
+            outgoingData->open(QIODevice::ReadOnly);
+
+            http_data += "Content-Length: ";
+            http_data += QByteArray::number(outgoingData->size());
+            http_data += "\r\n\r\n";
+            http_data += outgoingData->readAll();
+        }
+    }
+
+    http_data += "\r\n";
+
+    local_socket->write(http_data);
+    local_socket->flush();
+}
+
+void mp::LocalSocketReply::read_reply()
+{
+    local_socket->read(reply_data.data(), len);
+
+    parse_reply();
+
+    emit finished();
+}
```

Comment:
This assumes that the reply will never be > 64k long, right? Add a comment?

Reply:
For right now, yes.  It should really loop there until `-1` or `0` is returned from `read()`.  Perhaps I should go ahead and fix that.

Reply:
Done, accept I didn't add an explicit `-1` case, but instead wrapped it all in a `> 0` case to keep looping.